Ruby 2.7 warning: creating a Proc without a block

utilum opened the pull request:
As of https://github.com/ruby/ruby/commit/9f1fb0a17feb calls to
`Proc.new` without giving a block tickles:

```
warning: tried to create Proc object without a block
```

We have a few such calls in Rails. This commit fixes the easier instances.

See:
https://travis-ci.org/rails/rails/jobs/487205819#L1161-L1190
https://travis-ci.org/rails/rails/jobs/487205821#L1154-1159
https://travis-ci.org/rails/rails/jobs/487205821#L1160-L1169
https://travis-ci.org/rails/rails/jobs/487205821#L1189

Two additional instances remain to be fixed.

The first, in [`ActionView#form_with`](https://github.com/rails/rails/blob/master/actionview/lib/action_view/helpers/form_helper.rb#L755) may require changing a method signature that is already long and ends with a double splat, as well as many internal calls.

The second, in [`ActiveSupport::Fanout#subscribe'](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/notifications/fanout.rb#L21) just leaves me baffled :( It's a very popular method in the test suite, gets tickled over 10_000 times when I run the Rails tests.

utilum committed:
Ruby 2.7 warning: creating a Proc without a block

As of [Revision 66772](
https://bugs.ruby-lang.org/projects/ruby-trunk/repository/trunk/revisions/66772)
 `Proc.new` without giving a block emits `warning: tried to create Proc object without a block`.

This commit fixes cases where Rails test suit tickles this warning.

See CI logs:
https://travis-ci.org/rails/rails/jobs/487205819#L1161-L1190
https://travis-ci.org/rails/rails/jobs/487205821#L1154-1159
https://travis-ci.org/rails/rails/jobs/487205821#L1160-L1169
https://travis-ci.org/rails/rails/jobs/487205821#L1189
https://travis-ci.org/rails/rails/jobs/487254404#L1307-L1416
https://travis-ci.org/rails/rails/jobs/487254405#L1174-L1191

matthewd commented:
Thanks! Now that declaring &block is zero-cost unless it's used, it'll be nice to drop this quirky spelling.

---


`Fanout#subscribe` (and `StatementCache.create`) needs something like:

```ruby
def subscribe(pattern = nil, callable = nil, &block)
  subscriber = Subscribers.new(pattern, callable || block)
```

---

Adding `&block` to `form_with` seems fine. I don't see how the change would affect more than the line you've highlighted and the `def` above?

utilum commented:
Thank you !

Fixed the remaining two.

utilum commented:
@matthewd does `StatementCash.create` need `callable || block`? What is the case for which `def create(connection, &block)` does not work?

matthewd commented:
`StatementCache.create(some_conn, lambda { .. })` (or more likely, some non-proc object with a `call` method.

(I have no particular opinion on how important that API is to preserve, but the method currently accepts either one regular positional parameter and a block parameter _or_ two regular positional parameters -- if we just want to resolve the new warning without making a semantic change, that's what we need to continue supporting.)

tenderlove reviewed (approved):
https://bugs.ruby-lang.org/issues/15539

utilum committed:
ActionPack Proc.new without a block

This commit fixes cases that use pre Ruby
[r66772](https://bugs.ruby-lang.org/projects/ruby-trunk/repository/trunk/revisions/66772)
syntax that are not tickled by the test suite.

utilum commented:
Thank you all.
I've added fixes to the `ActionPack` methods.
Two testing questions:
- The `ActionPack` tests did not warn even before this change. Do we want to add tests?
- Do we want tests that verify the changed methods' behavior with both a positional callabale parameter and with a block?

rafaelfranca commented:
> * The `ActionPack` tests did not warn even before this change. Do we want to add tests?

Yes

> * Do we want tests that verify the changed methods' behavior with both a positional callabale parameter and with a block?

Yes

utilum commented:
Considering that we want to test for warnings, and that a very similar test would be applied in six components, I'd like to use 'warning' gem, and ideally add a couple of assertions for internal use to `ActiveSupport::Testing::Assertions`. Say `assert_warns` and `assert_not_warns`.

Overkill?
